agent/job_runner.go

```diff
@@ -283,30 +285,56 @@ func (r *JobRunner) Run() error {
 	signal := ""
 	signalReason := ""
 
-	// Run the process. This will block until it finishes.
-	if err := r.process.Run(); err != nil {
-		// Send the error as output
-		r.logStreamer.Process(fmt.Sprintf("%s", err))
+	// Before executing the bootstrap process with the received Job env,
+	// execute the pre-bootstrap hook (if present) for it to tell us
+	// whether it is happy to proceed.
+	environmentCommandOkay := true
 
-		// The process did not run at all, so make sure it fails
-		exitStatus = "-1"
-		signalReason = "process_run_error"
-	} else {
-		// Add the final output to the streamer
-		r.logStreamer.Process(r.output.String())
+	if hook, _ := hook.Find(r.conf.AgentConfiguration.HooksPath, "pre-bootstrap"); hook != "" {
+		// Once we have a hook any failure to run it MUST be fatal to the job to guarantee a true
+		// positive result from the hook
+		okay, err := r.executePreBootstrapHook(hook)
+		if !okay {
+			environmentCommandOkay = false
+
+			// Ensure the Job UI knows why this job resulted in failure
+			r.logStreamer.Process("pre-bootstrap hook rejected this job, see the buildkite-agent logs for more details")
```

Comment:
Made a tweak to this error message since the screenshot to point folks towards the agent jobs